### PR TITLE
Add citationType property on Cite nodes

### PR DESCRIPTION
**Pull request by nokome**
This adds the `citationType` property to `Cite` nodes. The `citationType` is an array of values from the `CitationTypeEnumeration` which is based on the [Citation Typing Ontology (CiTO)](http://www.sparontologies.net/ontologies/cito). The use cases for this property are:

- allowing authors to add CiTO annotations when creating citations, for example in Markdown, following the approach of the [Journal of Cheminformatics](https://github.com/jcheminform/markdown-jcheminf/blob/master/templates/research_article.Rmd) (see also [this pilot](https://www.biomedcentral.com/collections/cito)):

```markdown
If you have more than one intention to cite, separate them with a colon, for example [@uses_method_in:extends:Willighagen_2020].
```

- allowing enrichment of document, for example using AI, to suggest CiTO annotations based on preceding text.

Given that the CiTO annotations already [exist in a JSON-LD context](https://raw.githubusercontent.com/SPAROntologies/cito/b4c5b10a8d7b5f0da6ea8a4f3edcc00d7984f9a4/docs/current/cito.json), I have written a script to "import" them as a schema.org `Enumeration`. This allows the mapping between enumeration members and the original CiTO annotations to be preserved (using `@id`). That has required some extensions to how we process JSON Schema and generate language bindings.

I addition, this PR makes some improvements to the development workflow (testing; updating mappings between generated schemas and file types).

My main question for reviewers is whether we should have a separate property, perhaps named `citationDistance`, for the following CiTO annotations: `AffilationSelfCitation`, `AuthorNetworkSelfCitation`, `AuthorSelfCitation`, `DistantCitation`, `FunderSelfCitation`, `JournalCartelCitation`, `JournalSelfCitation`, `SelfCitation`, `SharesAuthorInstitutionWith`, `SharesAuthorWith`, `SharesFundingAgencyWith`, `SharesJournalWith`, `SharesPublicationVenueWith`. These all reflect the relation between those involved in the citing and the cited, rather than the nature or intent of the citation itself.

### CODE REVIEW

**Review by egonw — COMMENTED** [review]
Hi, thanks for doing this! Makes me very happy. 

Regarindg the citationDistance question I am not sure. This was not in the scope of our pilot and not something I would ask they author to add anyway. For things like self-citations, these can be computationally derived from the citation network in combination with ORCID. That is, I do not think you need those: these are qualifications when analyzing an article, not when authoring it.

**Review by rgieseke — APPROVED** [review]

**Comment by nokome** [comment]
Thanks for having a look over this @egonw.

> That is, I do not think you need those: these are qualifications when analyzing an article, not when authoring it.

Agreed, I will drop the `*Citation` and `Shares*` annotations from the enumeration. We can add a `citationDistance` property with those members if / when we have a use case for them (e.g. doing analysis on an article).

File: schema/Cite.schema.yaml [hunk 1]
```diff
@@ -37,6 +37,12 @@ properties:
       LaTeX's `natbib` package has `\citep{}` (for `parenthetical`), `\citet{}` (for `textual`),
       `\citeauthor{}` (for `textual-author`), `\citeyear{}` (for `textual-year`).
       See https://www.overleaf.com/learn/latex/Natbib_citation_styles.
+  citationType:
```
Comment (alex-ketch) [inline schema/Cite.schema.yaml:40]: Just a small suggestion, but what do you think about naming this `citationIntent` to clarify the purpose?
Personally I find nuance between `citationMode` and `citationType` harder to reason about.

**Review by alex-ketch — APPROVED** [review]
Looks good @nokome, just a small naming suggestion.

File: schema/Cite.schema.yaml (hunk 1 above)
Comment (nokome) [reply schema/Cite.schema.yaml:40]: I think that is a great suggestion!

**Comment by egonw** [comment]
Sounds awesome. Now, I am not familiar with the code base, and while I see what it was doing, I am not sure what user-visible changes this means. That is, I did not spot obvious UI hooks. I guess this was just the foundation?

**Comment by nokome** [comment]
Yes, this is the foundational data layer. By adding `citationIntent` here, to the data schema, we can target it when parsing Markdown and other formats in https://github.com/stencila/encoda

**Comment by egonw** [comment]
thanks! well, I'm very excited about it :)

**Comment by egonw** [comment]
update from our side, I've blogged about our progress here: https://blogs.biomedcentral.com/on-physicalsciences/2021/07/20/typed-citations-journal-of-cheminformatics/

**Pull request merged**